### ISSUE TITLE
Add LHS non-null assert semantics to operator assignments

### ISSUE THREAD

**Issue by hawketh** [Microsoft/TypeScript#16064]
**TypeScript Version:**  2.2.1 / nightly (2.2.0-dev.201xxxxx)

**Code**
```ts
interface A { field: number|null; }
let item: A = {field: 0};
item.field += 1;
```

**Expected behavior:**
As written, it should raise an error (which it does), because `item.field` could be null.  What I'd like to be able to do is use the assertion syntax `item.field! += 1;` so that I can let the compiler know that the LHS (lvalue of the operator assignment) is not null or undefined.  This is according to the usual semantics that an operator assignment can be interpreted as equivalent to `lvalue = lvalue op rvalue;` (with singular occurrence of lvalue side-effects).

This syntax is particularly useful where A has been initialized with a non-null value for the field.  While in this example the inference engine might be able to identify that `field` is not null, this won't work if you have parameterized constructors, such as:
```ts
class B { field: number|null; constructor(...){ ... } }
let item = new B({field: 0});
```
You can't infer that the constructor did, in fact, set a value into `field`.

**Actual behavior:**
The compiler complains about the extended syntax.  That is, it gets confused with the assertion and the operator assignment.  You're forced to use the expanded form making this somewhat less readable.

**Comment by mhegazy** [comment]
looks like a duplicate of https://github.com/Microsoft/TypeScript/issues/15796

**Comment by hawketh** [comment]
The description in [#15796](https://github.com/Microsoft/TypeScript/issues/15796) would cover this as long as you generalize that example to any lvalue expression.  Operator assignment and post-access side effect operators both require lvalues, and it should be possible to assert that the lvalue won't be null.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
